File: 32-longest-valid-parentheses/longest-valid-parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        vector<int> v;
        v.push_back(-1);
        int max_length = 0;

        for (int idx = 0; idx < s.length(); idx++) {
            if (s[idx] == '(') {
                v.push_back(idx);
            } else {
                v.pop_back();
                if (v.empty()) {
                    v.push_back(idx);
                } else {
                    max_length = max(max_length, idx - v.back());
                }
            }
        }

        return max_length;
    }
};

File: 32-longest-valid-parentheses/longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack = [-1]
        max_length = 0

        for idx, char in enumerate(s):
            if char == "(":
                stack.append(idx)
            else:
                stack.pop()
                if len(stack) != 0:
                    max_length = max(max_length, idx - stack[-1])
                else:
                    stack.append(idx)

        return max_length
